apps/prairielearn/src/pages/studentInstanceQuestion/studentInstanceQuestion.js

```diff
@@ -280,10 +283,34 @@ router.get(
 router.get(
   '/',
   asyncHandler(async (req, res) => {
-    const variant_id =
+    let variant_id =
       res.locals.assessment.type === 'Exam' || typeof req.query.variant_id !== 'string'
         ? null
         : req.query.variant_id;
+
+    const isAssessmentAvailable =
+      _.get(res.locals, 'assessment_instance.open', true) &&
+      _.get(res.locals, 'authz_result.active', true);
+
+    if (variant_id === null && !isAssessmentAvailable) {
+      // We can't generate a new variant in this case, so we
+      // fetch and display the most recent non-broken variant.
+      // If no such variant exists, we tell the user that a new variant
+      // cannot be generated.
+      const result = await sqldb.queryZeroOrOneRowAsync(sql.select_last_variant_id, {
+        instance_question_id: res.locals.instance_question.id,
+      });
```

Comment:
We have been moving towards newer Zod-validated queries in new code.
```suggestion
      const last_variant_id = await sqldb.queryOptionalRow(
        sql.select_last_variant_id,
        { instance_question_id: res.locals.instance_question.id },
        IdSchema,
      });
```
You will need to import `IdSchema` from lib/db-types.ts, and change the code after this point to use `last_variant_id` instead of `result`. `last_variant_id` will be null if it does not exist.

Reply:
Done